src/Microsoft.AspNetCore.WebHooks.Receivers/Filters/WebHookVerifySignatureFilter.cs

```diff
@@ -192,19 +227,101 @@ protected static bool SecretEqual(byte[] inputA, byte[] inputB)
+        /// </param>
+        /// <returns>
+        /// A <see cref="Task"/> that on completion provides a <see cref="byte"/> array containing the SHA1 HMAC of
+        /// the <paramref name="prefix"/> followed by the <paramref name="request"/>'s body.
+        /// </returns>
+        protected Task<byte[]> GetRequestBodyHash_SHA1(HttpRequest request, byte[] secret, byte[] prefix)
+        {
+            return GetRequestBodyHash_SHA1(request, secret, prefix, suffix: null);
+        }
+
+        /// <summary>
+        /// Returns the SHA1 HMAC of the given <paramref name="prefix"/>, the given <paramref name="request"/>'s
+        /// body, and the given <paramref name="suffix"/> (in that order).
+        /// </summary>
+        /// <param name="request">The current <see cref="HttpRequest"/>.</param>
+        /// <param name="secret">The key data used to initialize the <see cref="HMACSHA1"/>.</param>
+        /// <param name="prefix">
+        /// If non-<see langword="null"/> and non-empty, additional <c>byte</c>s to include in the hashed content
+        /// before the <paramref name="request"/>'s body.
+        /// </param>
+        /// <param name="suffix">
+        /// If non-<see langword="null"/> and non-empty, additional <c>byte</c>s to include in the hashed content
+        /// after the <paramref name="request"/>'s body.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Task"/> that on completion provides a <see cref="byte"/> array containing the SHA1 HMAC of
+        /// the <paramref name="prefix"/>, the <paramref name="request"/>'s body, and the <paramref name="suffix"/>
+        /// (in that order).
         /// </returns>
-        protected virtual async Task<byte[]> GetRequestBodyHash_SHA1(HttpRequest request, byte[] secret)
+        protected virtual async Task<byte[]> GetRequestBodyHash_SHA1(
+            HttpRequest request,
+            byte[] secret,
+            byte[] prefix,
+            byte[] suffix)
         {
             await WebHookHttpRequestUtilities.PrepareRequestBody(request);
 
             using (var hasher = new HMACSHA1(secret))
             {
                 try
                 {
-                    var hash = hasher.ComputeHash(request.Body);
-                    return hash;
+                    if (prefix != null && prefix.Length > 0)
+                    {
+                        hasher.TransformBlock(
+                            prefix,
+                            inputOffset: 0,
+                            inputCount: prefix.Length,
+                            outputBuffer: null,
+                            outputOffset: 0);
+                    }
+
+                    // Split body into 4K chunks.
+                    var buffer = new byte[4096];
+                    var inputStream = request.Body;
+                    int bytesRead;
+                    while ((bytesRead = inputStream.Read(buffer, 0, buffer.Length)) > 0)
```

Comment:
Assuming these are synchronous reads because we've already buffered the body?

Reply:
More like a typo. It's actually fine either way but `await ...ReadAsync(...)` avoids questions.